### ISSUE TITLE
Make ambiguous field accesses an error

### ISSUE THREAD

**Issue by ngoldbaum** [yt-project/yt#1467]
Right now yt lets you use a string field name instead of a full field tuple to access a field. To translate from the field name to a field tuple it uses a set of heuristics to guess the field type corresponding to the field name.

Unfortunately, sometimes it guesses wrong, or worse, the guess changes based on IO patterns. For example:

```
In [8]: ds._get_field_info('particle_mass')
Out[8]: Alias Field for "('PartType0', 'Masses')" (PartType0, particle_mass): (units: code_mass, particle field)

In [9]: ad['all', 'particle_mass']
Out[9]:
YTArray([  8.93692344e+39,   1.07352366e+40,   9.48756720e+39, ...,
           8.89038231e+39,   8.87535665e+39,   8.91641743e+39]) g

In [10]: ds._get_field_info('particle_mass')
Out[10]: Alias Field for "('all', 'Masses')" (all, particle_mass): (units: g, particle field)
```

(`ds._get_field_info` is a private function used to facilitate this field guessing system)

I'd like to propose a backward incompatible change. Rather than doing this guessing business and sometimes being wrong or sometimes switching which field we return based on the history of field accesses, we should instead unconditionally raise an error if a field access is ambiguous.

Concurrently, we should also update our documentation to use either field tuples or access fields using the `ds.fields` object,  (like `ds.fields.gas.density` instead of `('gas', 'density')`).

I'm not sure how disruptive this is but I bet there are a bunch of places inside of yt where we are doing ambiguous field accesses.

**Comment by qobilidop** [comment]
Could we also add a feature to temporarily set a default field type to work with? Because it's very common to do analysis on a single field type. And it would be tedious and error-prone having to type it explicitly every time.

I'm proposing a context manager API:

```
with ad.default_field_type('all'):
    print(ad['particle_mass'])
```

And except for within the context, field access using the short-hand string form would err.

**Comment by chummels** [comment]
That's an interesting idea. So inside the context manager, you could use strings as fields or also tuples of fields with other ftypes, but outside the context manager, you could only use tuples as fields?

That seems pretty useful for shorthanding.

**Comment by chummels** [comment]
This topic was just discussed at a yt-4.0 release meeting. It was agreed that in yt-4.0, we will accept both string and tuple field names. If someone specifies a string field name and there is any ambiguity if there are two potential tuple fields that could be accessed, the code will error out with an error message requiring the user to be more explicit.

The issue was raised that it might be confusing to some users that certain fields differ only by capitalization (e.g., "Density" and "density" for the Enzo frontend: ("Enzo", "Density") and ("gas", "density")), but the consensus was to remain case-sensitive, and just include a note in the Field documentation that this was the expected behavior.

**Comment by cphyc** [comment]
I just thought that maybe we would want to raise a warning when the ambiguity comes from case-insensitive cases, so that the user may realize there is a difference between `"Density"` and `"density"`?

**Comment by chummels** [comment]
I think that's reasonable to issue a non-failing warning.  So if I'm using an Enzo dataset with ("Enzo", "Density") and ("gas", "density") and I run:

```
import yt
yt.load('IsolatedGalaxy/galaxy0030/galaxy0030')
yt.SlicePlot(ds, 'z', 'density').save()
```

yt correctly uses the ("gas", "density") field, runs the command, and returns a warning:

```
WARNING: Dataset ds has both ("Enzo", "Density") and ("gas", "density") fields. SlicePlot using ("gas", "density").
```

Or something like that.

**Comment by chummels** [comment]
Note, I am closing Issue #832 because it is a repeat of this issue.

**Comment by cphyc** [comment]
Should we consider this as closed or not? At the moment, we throw a deprecation warning (rather than an error) whenever an ambiguous field is accessed.

I'd be in favour of closing :)

**Comment by jzuhone** [comment]
I also agree to close. I'll do it now.

**Comment by jzuhone** [comment]
Closed by PR #2967
